db/src/Pos/DB/Block/Internal.hs

```diff
@@ -0,0 +1,290 @@
+import           System.FilePath ((</>))
+import           System.IO (IOMode (WriteMode), hClose, hFlush, openBinaryFile)
+import           System.IO.Error (IOError, isDoesNotExistError)
+
+import           Pos.Binary.Class (decodeFull', serialize')
+import           Pos.Chain.Block (Block, GenesisBlock, HeaderHash,
+                     SlogUndo (..), Undo (..), headerHash)
+import qualified Pos.Chain.Block as CB
+import           Pos.Chain.Delegation (DlgUndo (..))
+import           Pos.Crypto (hashHexF)
+import           Pos.DB.BlockIndex (deleteHeaderIndex, putHeadersIndex)
+import           Pos.DB.Class (MonadDB (..), MonadDBRead (..), Serialized (..),
+                     SerializedBlock, SerializedBlund, SerializedUndo)
+import           Pos.DB.Error (DBError (..))
+import           Pos.DB.Pure (DBPureVar, MonadPureDB, atomicModifyIORefPure,
+                     pureBlocksStorage)
+import           Pos.DB.Rocks (MonadRealDB, blockDataDir, getNodeDBs)
+import           Pos.Util.Util (HasLens (..), eitherToThrow)
+
+
+----------------------------------------------------------------------------
+-- Implementations for 'MonadRealDB'
+----------------------------------------------------------------------------
+
+-- Get serialization of a block with given hash from Block DB.
+getSerializedBlock
+    :: MonadRealDB ctx m
+    => HeaderHash -> m (Maybe SerializedBlock)
+getSerializedBlock hh =
+    Serialized . fst <<$>> getSerializedBlund hh
+
+-- Get serialization of an undo data for block with given hash from Block DB.
+getSerializedUndo
+    :: MonadRealDB ctx m
+    => HeaderHash -> m (Maybe SerializedUndo)
+getSerializedUndo hh =
+    Serialized . snd <<$>> getSerializedBlund hh
+
+-- Get serialization of a block with given hash from Block DB.
+getSerializedBlund
+    :: MonadRealDB ctx m
+    => HeaderHash -> m (Maybe (ByteString, ByteString))
+getSerializedBlund hh = do
+    bsp <- flip getAllPaths hh . view blockDataDir <$> getNodeDBs
+    blundExists <- liftIO $ doesFileExist (bspBlund bsp)
+    if blundExists
+    then do
+      mbs <- getRawData $ bspBlund bsp
+      case mbs of
+        Nothing -> pure Nothing
+        Just ser -> eitherToThrow . bimap DBMalformed Just
+                    $ decodeFull' @(ByteString, ByteString) ser
+    else consolidateBlund hh
+
+-- | Read independent block and undo data and consolidate them into a single
+-- blund file.
+consolidateBlund
+    :: MonadRealDB ctx m
+    => HeaderHash
+    -> m (Maybe (ByteString, ByteString))
+consolidateBlund hh = do
+    bsp <- flip getAllPaths hh . view blockDataDir <$> getNodeDBs
+    block <- getRawData $ bspBlock bsp
+    undo <- getRawData $ bspUndo bsp
+    case (,) <$> block <*> undo of
+        Just blund -> do
+            putRawData (bspBlund bsp) $ serialize' blund
+            liftIO . removeFile $ bspBlock bsp
+            liftIO . removeFile $ bspUndo bsp
+            return $ Just blund
+        Nothing -> return Nothing
```

Comment:
Might be clearer to use `Maybe`'s traversable instance:

```haskell
for ((,) <$> block <*> undo) $ \blund -> do
    -- ...
    return blund
```

Reply:
Also code moved from another file, but i will fix it.